src/librustc_codegen_llvm/builder.rs

```diff
@@ -153,35 +146,46 @@ impl BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
         unsafe {
             llvm::LLVMPositionBuilderBefore(self.llbuilder, instr);
         }
-
-        let pos = unsafe { llvm::LLVMRustInstructionIndex(instr) };
-
-        let bb = unsafe { llvm::LLVMGetInstructionParent(instr) } as *const llvm::BasicBlock
-            as *const sir::BasicBlock;
-
-        let builder = self.llbuilder as *const llvm::Builder<'_> as *const sir::Builder;
-
-        self.cx.with_sir_cx_mut(|sir_cx| {
-            sir_cx.position_before(builder, bb, pos);
-        });
     }
 
     fn position_at_end(&mut self, llbb: &'ll BasicBlock) {
         unsafe {
             llvm::LLVMPositionBuilderAtEnd(self.llbuilder, llbb);
         }
-        let builder = self.llbuilder as *const llvm::Builder<'_> as *const sir::Builder;
-        let bb = llbb as *const llvm::BasicBlock as *const sir::BasicBlock;
-        self.cx.with_sir_cx_mut(|sir_cx| {
-            sir_cx.position_at_end(builder, bb);
-        });
     }
 
-    fn add_yk_block_label(&mut self, block: &'ll BasicBlock, lbl_name: CString) {
-        let dbg_cx = self.cx().dbg_cx.as_ref().unwrap();
-        let di_bldr = dbg_cx.get_builder();
-        unsafe {
-            llvm::LLVMRustAddYkBlockLabel(self.llbuilder, di_bldr, block, lbl_name.as_ptr());
+    fn add_yk_block_label(&mut self, fname: &str, sym: &SymbolName, bbidx: usize) {
+        if !self.tcx.sess.opts.cg.tracer.sir_labels() {
+            // We are not in hardware tracing mode.
+            return;
+        }
+
+        if self.tcx.crate_name(LOCAL_CRATE).as_str().starts_with("rustc") {
+            // Skip rustc crates since we will never trace them.
+            return;
+        }
+        if fname == "core::intrinsics::drop_in_place"
+            || fname == "std::intrinsics::drop_in_place"
+            || fname == "ptr::drop_in_place"
+            || sym.name.as_str() == "main"
+        {
+            // Generating labels for these functions results in segfaults.
+            return;
+        }
+
+        // Create the label
+        let lbl_name = CString::new(format!("{}:{}:{}", BLOCK_LABEL_PREFIX, sym, bbidx)).unwrap();
+
+        if let Some(dbg_cx) = self.cx().dbg_cx.as_ref() {
+            let di_bldr = dbg_cx.get_builder();
+            unsafe {
+                llvm::LLVMRustAddYkBlockLabel(
+                    self.llbuilder,
+                    di_bldr,
+                    self.llbb(),
+                    lbl_name.as_ptr(),
```

Comment:
Does `LLVMRustAddYkBlockLabel` copy the contents pointed to, or do we need to leak `lbl_name` or ...?

Reply:
It's a good question.

The pointer gets passed directly into LLVM's `DIBuilder::createLabel()`. I tried to follow the call down to see what happens, but I quickly got lost in a tangle of C++.

Lots of LLVM API functions accept `char *` arguments for names of things. I'd hope that the API takes a copy, or the API would be kind of absurd (?). The user would have to ensure that the strings stay alive.

What do you think?

Reply:
I think we need to be sure about this, otherwise we're going to create code that's likely to be either wrong or inefficient.

Reply:
OK. After trawling the llvm sources some, I've figured it out.

[](url)

LLVM uses [StringMaps](https://llvm.org/doxygen/classllvm_1_1StringMap.html) to cache strings.

When we  do `di_builder->createLabel(..., labelName)`, the name is eventually passed to a `StringMap` responsible for caching metadata (debug info) strings. [try_emplace()](https://llvm.org/doxygen/classllvm_1_1StringMap.html#a3f1b153d947276395158efe9e5aa7693) is called and if there's a cache miss, the [string is memcpy()d into the cache](https://llvm.org/doxygen/StringMap_8h_source.html#l00197).

If the entry is already in the cache, the old value is used.

Since the string is copied into the cache, we don't have to ensure that strings in our arguments to API calls survive.

Reply:
OK, cool!